source/server/listener_impl.cc

```diff
@@ -26,19 +26,36 @@
 namespace Envoy {
 namespace Server {
 
-ListenSocketFactoryImplBase::ListenSocketFactoryImplBase(
-    ListenerComponentFactory& factory, Network::Address::InstanceConstSharedPtr local_address,
-    Network::Address::SocketType socket_type, const Network::Socket::OptionsSharedPtr& options,
-    bool bind_to_port, const std::string& listener_name)
-    : factory_(factory), local_address_(local_address), socket_type_(socket_type),
-      options_(options), bind_to_port_(bind_to_port), listener_name_(listener_name) {}
-
-Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApplyOptions() {
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(ListenerComponentFactory& factory,
+                                                 Network::Address::InstanceConstSharedPtr address,
+                                                 Network::Address::SocketType socket_type,
+                                                 const Network::Socket::OptionsSharedPtr& options,
+                                                 bool bind_to_port,
+                                                 const std::string& listener_name, bool reuse_port)
+    : factory_(factory), local_address_(address), socket_type_(socket_type), options_(options),
+      bind_to_port_(bind_to_port), listener_name_(listener_name), reuse_port_(reuse_port) {
+
+  // If reuse_port_ is false, create a socket here which will be used by all worker
+  // threads; otherwise, if port is 0, still need to create a socket here, because
+  // a random port will be assigned, then all worker threads should use same port.
+  if (!reuse_port_ || local_address_->ip()->port() == 0) {
```

Comment:
I would like to not create a temporary socket just to get ephemeral port number if possible. Can you try the approach I proposed in my other comment to see if it's possible?

Reply:
on worker thread things happen currently, a real port is reserved with bind() for port 0, I guess there are only two ways to resolve this problem:
- main thread bind port 0 to reserve a port
- worker threads do mutex to guarantee the first one can reserve a port an others use same

We'd better avoid using mutex for reserving a port.

Reply:
This discussion continues in another comments ;)